Feature sending mentorship request.

DeemaP opened the pull request:
add new Exceptions
add new Validators
add MentorshipRequestController, MentorshipRequestService, MentorshipRequestDto and tests.

DeemaP committed:
Feature sending mentorship request.
add new Exceptions
add new Validators
add MentorshipRequestController, MentorshipRequestService, MentorshipRequestDto and tests.

drewbal commented on src/main/java/school/faang/user_service/controller/MentorshipRequestController.java:
```diff
@@ -0,0 +1,24 @@
+        if (dto.getDescription() == null || dto.getDescription().isBlank()) {
+            log.warn("Mentorship request failed: description is missing or blank.");
+            throw new InvalidMentorshipRequestException("Description is missing or blank.");
+        }
```
Может валидаторам - валидаторово, а контроллерам - контроллерово? :)

DeemaP replied:
Наверное да, почему-то я тупую проверку тут оставил и все) Надо перекинуть в валидатор)

drewbal commented on src/main/java/school/faang/user_service/controller/MentorshipRequestController.java:
```diff
@@ -0,0 +1,24 @@
+        }
+        log.info("Validation for null or blank description successful.");
+        requestService.requestMentorship(dto);
+    }
```
Ну вот чисто погундеть. А может все же что-то возвращать обратно (хотя в задании четко не обозначено). Просто по логике юзер присылает запрос на эндпойнт, ждет какого-то ответа, ну как минимум будет знать id своего запроса, когда будет звонить в индийский колл-центр и узнавать, где ментор.
Но это всегда можно будет прикрутить и в будущем, но с работой в обе стороны твой контроллер сможет тусить с другими крутыми парнями.

DeemaP replied:
Я дальше по заданию это поправлю) Специально пока не додумывал лишнюю логику, иду по пунктам и пилю то, что нужно конкретно сейчас))

RuslanZinovyev replied:
Учитывая что это контроллер лучше все же вернуть что-то, хотя бы информацию об успешном запросе с кодом 200.
Сделай возвращаемым значением ResponseEntity<Void> с типом Void и верни return ResponseEntity.ok().build();

DeemaP replied:
Сделал сами dto возвращаемыми значениями, чтоб было как у всех) Дальше по проекту будем плясать по ходу дела)

RuslanZinovyev replied:
Я очень рекомендую пользоваться ResponseEntity для контроллера, но возможно на данном этапе для тебя не совсем очевидны преимущества и удобства данного класса, поэтому разберемся с этим позже.

DeemaP replied:
На досуге почитаю как это работает и постараюсь внедрить. Спасибо)

drewbal commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,24 @@
+
+    public void requestMentorship(MentorshipRequestDto dto) {
+        if (requestValidator.validateMentorshipRequest(dto)) {
+            log.info("Request dto validation successful. Creating request.");
```
ну раз начал про логи, то и привесь в него что-нибудь, типа id. А то что нам искать потом в одинаковых фразах?

DeemaP replied:
Я вообще не уверен, что тут нужно такое вот логирование, но для информативности согласен, надо докинуть айдишку хотя бы)

drewbal commented on src/main/java/school/faang/user_service/validator/MentorshipRequestValidator.java:
```diff
@@ -0,0 +1,59 @@
+package school.faang.user_service.validator;
```
Этот класс - мощь!

DeemaP committed:
Apply review feedback:
unify validation method names
remove return values from validation methods
refactor tests to match changes

DeemaP committed:
Feature - get all mentorship requests.
add getRequests() to Controller and Service
add filters
add additional validation
add tests for new methods and classes

DeemaP committed:
Feature - accept mentorship request.
add acceptRequest() to Controller and Service
add additional validation
add tests for new methods and classes
refactored exceptions and tests

DeemaP committed:
Feature - reject mentorship request.
add rejectRequest() to Controller and Service
add additional validation
add tests for new methods and classes
refactored exceptions and tests

DeemaP committed:
Merge branch 'refs/heads/Dragon-master-stream7' into feature-BJS2-40489

# Conflicts:
#	src/main/java/school/faang/user_service/exception/EntityNotFoundException.java

RuslanZinovyev commented on src/main/java/school/faang/user_service/filter/MentorshipRequestFilter/DescriptionFilter.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Override
+    public Stream<MentorshipRequest> apply(Stream<MentorshipRequest> requests, RequestFilterDto filter) {
+        return requests.filter(request -> request.getDescription().toLowerCase()
```
Ты можешь использовать метод класса String equalsIgnoreCase() вместо toLowerCase() что более дорогая процедура.

DeemaP replied:
Смысл в том, что я фильтрую не по полному совпадению описания, а по какому-то слову, содержащемуся в нем, поэтому делаю toLowerCase(), чтобы избежать необходимости учитывать регистр при вводе фильтра.

Ведь условный "I neeh help with smth".equalsIgonreCase("HELP") не выдаст true,

RuslanZinovyev replied:
Я по моему оставил комментарий не на том методе, что до этого видел =)
В этом ты все правильно сделал.

RuslanZinovyev replied:
Ok

DeemaP committed:
Add logging.
Update controller and service return values.

mighty-mhsl commented on src/main/java/school/faang/user_service/filter/MentorshipRequestFilter/RequestFilter.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.util.stream.Stream;
+
+public interface RequestFilter {
```
А нет такого, что дальше ребята под другие сущности будут писать такой же интерфес, но он будет просто другие типы данных принимать? Может сделаем его дженерик, чтобы потом все его для других фильтров могли использовать?

DeemaP replied:
Справедливо! Сделал дженерик интерфейс и перенес в общий пакет Filter

mighty-mhsl commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,82 @@
+    private final MentorshipRequestValidator requestValidator;
+    private final MentorshipRequestMapper requestMapper;
+    private final List<RequestFilter> requestFilters;
+    private MentorshipRequest request;
```
А это зачем тут полем? Надо удалить
Не могут наши классы хранить состояние в памяти - мы все только в БД храним

DeemaP replied:
Применял в двух методах, по старинке решил вынести в общую переменную класса) Убрал в методы

mighty-mhsl commented on src/main/java/school/faang/user_service/service/MentorshipRequestService.java:
```diff
@@ -0,0 +1,82 @@
+
+        return requestRepository.findById(id).orElseThrow(() -> {
+            log.warn("Request with id '{}' is Null", id);
+            return new EntityNotFoundException("Request with id '" + id + "' is Null");
```
Он скорее в БД не найден, а не null, можно более точное сообщение указать

DeemaP replied:
Поправил

mighty-mhsl commented on src/main/java/school/faang/user_service/validator/MentorshipRequestValidator.java:
```diff
@@ -0,0 +1,101 @@
+    }
+
+    private LocalDateTime getThreeMonthAgo() {
+        return LocalDateTime.now(ZoneId.of("UTC")).minusMonths(3);
```
Что такое 3? Может в константу?

DeemaP replied:
Вынес в константу и изменил названия методов, чтоб соответствовали)

mighty-mhsl commented on src/main/java/school/faang/user_service/validator/RequestFilterValidator.java:
```diff
@@ -0,0 +1,21 @@
+            log.warn("Filter cannot be null.");
+            throw new InvalidRequestFilterException("Filter cannot be null.");
+        }
+        log.info("Filters {} are not null.", filters);
```
Ну в инфо жестко такое логировать. Это в принципе обычная ситуация, так что лог такой не очень полезен, больше засорять будет нам

mighty-mhsl commented on src/test/java/school/faang/user_service/TestDataCreator.java:
```diff
@@ -0,0 +1,67 @@
+import school.faang.user_service.entity.RequestStatus;
+import school.faang.user_service.entity.User;
+
+public class TestDataCreator {
```
А зачем нам такой класс? Почему мы прямо в тестах все это не делаем?
Выглядит так, что в этот класс постепенно начнут сваливать все подряд, т.к. у него нет четкого назначения.  А это нехорошо, т.к. в нем потом не сориентируешься
Я считаю, что все данные нужно создавать в самих тестах или их родительских классах, если есть что-то общее

mighty-mhsl replied:
А этот удалить - он всем даст возможность в нем сделать мусорку постепенно

mighty-mhsl commented on src/test/java/school/faang/user_service/controller/MentorshipRequestControllerTest.java:
```diff
@@ -0,0 +1,74 @@
+    void testControllerAcceptRequest() {
+        requestController.acceptRequest(requestDto.getId());
+
+        verify(requestService, times(1)).acceptRequest(requestDto.getId());
```
А почему мы не делаем ассерт, что нам нужный id вернулся?

mighty-mhsl commented on src/test/java/school/faang/user_service/controller/MentorshipRequestControllerTest.java:
```diff
@@ -0,0 +1,74 @@
+    void testControllerRejectRequest() {
+        requestController.rejectRequest(requestDto.getId(), rejectionDto);
+
+        verify(requestService, times(1)).rejectRequest(requestDto.getId(), rejectionDto);
```
А тут?

mighty-mhsl commented on src/test/java/school/faang/user_service/controller/MentorshipRequestControllerTest.java:
```diff
@@ -0,0 +1,74 @@
+    void testControllerGettingRequests() {
+        requestController.getRequests(filterDto);
+
+        verify(requestService, times(1)).getRequests(any(RequestFilterDto.class));
```
А почему мы не проверяем, что нам getRequests вернул действительно нужный список с нужными данными?

mighty-mhsl commented on src/test/java/school/faang/user_service/filter/MentorshipRequestFilter/RequestFilterTest.java:
```diff
@@ -0,0 +1,110 @@
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+class RequestFilterTest {
```
Не, нужно писать отдельный тестовый класс под каждый отдельный класс фильтра. В этом и суть SOLID тоже, что у каждого - отдельное ВСЕ

mighty-mhsl commented on src/test/java/school/faang/user_service/service/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,157 @@
+        requestValidator.validateMentorshipRequest(requestDto);
+        requestService.requestMentorship(requestDto);
+
+        verify(requestRepository).create(requestDto.getRequesterId(),
```
Почему не делаем ассерт возвращаемого значения здесь?

mighty-mhsl commented on src/test/java/school/faang/user_service/service/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,157 @@
+
+        List<MentorshipRequestDto> requestDtoList = requestService.getRequests(filterDto);
+
+        assertEquals(1, requestDtoList.size());
```
А наполнение проверить?
Размер может быть норм, но наполенение может быть неверным

mighty-mhsl commented on src/test/java/school/faang/user_service/service/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,157 @@
+        assertThrows(EntityNotFoundException.class, () -> requestService.acceptRequest(firstRequestId));
+        verify(requestValidator, times(1)).validateMentorshipRequestExists(firstRequestId);
+        verify(requestRepository, times(1)).findById(firstRequestId);
+        verify(firstRequest, times(0)).getRequester();
```
Вот эти проверки уже слишком. Достаточно просто проверить вызовы зависимостей, а не вообще все возможные методы
Все, что ниже этого коммента - лишнее

mighty-mhsl commented on src/test/java/school/faang/user_service/service/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,157 @@
+        verify(firstRequest, times(1)).getReceiver();
+        verify(requestValidator, times(1)).validateRequesterHasReceiverAsMentor(any(), any());
+        verify(firstUser, times(1)).getMentors();
+        verify(firstRequest, times(1)).setStatus(RequestStatus.ACCEPTED);
```
И тут
Мы должны не факт вызовов методов проверять. А что состояние изменилось
Такие вещи мы должны проверять через ассерты. Если метод вызвался, это еще не значит, что значение поменялось
Поэтому не имеет смысла такие вызовы проверять. Надо просто проверить, что там внутри везде теперь правильные данные лежат

mighty-mhsl commented on src/test/java/school/faang/user_service/service/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,157 @@
+        assertThrows(EntityNotFoundException.class,
+                () -> requestService.rejectRequest(firstRequestId, rejectionDto));
+        verify(firstRequest, times(0)).setStatus(RequestStatus.REJECTED);
+        verify(firstRequest, times(0)).setRejectionReason(rejectionDto.getReason());
```
Ассерты, вместо этого

mighty-mhsl commented on src/test/java/school/faang/user_service/service/MentorshipRequestServiceTest.java:
```diff
@@ -0,0 +1,157 @@
+
+        assertDoesNotThrow(() -> requestService.rejectRequest(firstRequestId, rejectionDto));
+        verify(firstRequest, times(1)).setStatus(RequestStatus.REJECTED);
+        verify(firstRequest, times(1)).setRejectionReason(rejectionDto.getReason());
```
Ассерты

mighty-mhsl commented on src/test/java/school/faang/user_service/validator/RequestFilterValidatorTest.java:
```diff
@@ -0,0 +1,30 @@
+
+    @BeforeEach
+    void setUp() {
+        filter = null;
```
А это зачем? Он же и так null

DeemaP committed:
Make changes according to reviewer's comments.

Refactor validation for null and blank values
Refactor tests
Refactor rejectMentorshipRequest() and acceptMentorshipRequest() methods

DeemaP committed:
Merge branch 'refs/heads/Dragon-master-stream7' into feature-BJS2-40489

# Conflicts:
#	src/main/java/school/faang/user_service/dto/UserDto.java
#	src/main/java/school/faang/user_service/mapper/UserMapper.java
#	src/main/java/school/faang/user_service/service/UserService.java

DeemaP committed:
Conflicts resolve.

RuslanZinovyev commented on src/main/java/school/faang/user_service/controller/MentorshipRequestController.java:
```diff
@@ -0,0 +1,41 @@
+
+@Slf4j
+@Component
+@RequiredArgsConstructor
```
Я тут подумал,  а давай поднимем твой скилл на новый уровень и ты напишешь мне настоящий REST controller со всеми веселыми аннотациями для restful ендпоинтов как GetMapping PostMapping etc
Таким образом мы также удалим валидатор RequestFilterValidator из контроллера и перенесем его в сервис, на сам же контроллер мы повесим аннотацию Validated и будем валидировать все DTO декларативно, тем более что все нужные констрейнтс на твоих DTO уже висят. Если не понимаешь как это сделать посмотри код Андрея и попроси помощи у ребят или у меня.

DeemaP replied:
Поизучал, пообщался с Андреем, оказалось не так все страшно и сложно) Можно проверять)

DeemaP committed:
Feature - @RestController MentorshipRequestController

DeemaP committed:
Filter interface
refactor data naming.

drewbal commented on src/main/java/school/faang/user_service/controller/MentorshipRequestController.java:
```diff
@@ -0,0 +1,56 @@
+        return ResponseEntity.ok().body(requestService.getRequests(filters));
+    }
+
+    @PutMapping
```
Вот тут надо путь указать. Дело в том, что у тебя 2 метода с запросом PUT ломятся в один url `api/mentorshipRequests`. И не понятно, что им делать. 
1) Надо прописать путь для них в формате `@PutMapping(/users/accepts/{id}` (над дизайном пути можно подумать, но не суть). В данном случае у тебя запрос пойдет по пути `api -> mentorshipRequests -> users -> accepts` и применит переменную для операции. 
2) Для того, чтобы переменная из запроса передалась в метод используй аннотацию `@PathVariable`, т.е. `acceptRequest(@PathVariable @NotNull @Positive Long id)`

DeemaP replied:
Вот это спасибо)) Я не подумал о том, что ему важно по какому пути идти)

drewbal commented on src/main/java/school/faang/user_service/controller/MentorshipRequestController.java:
```diff
@@ -0,0 +1,56 @@
+
+@Slf4j
+@RestController
+@RequestMapping("api/mentorshipRequests")
```
Для url используй kebab-case, т.е. `mentorship-requests`

drewbal commented on src/main/java/school/faang/user_service/controller/MentorshipRequestController.java:
```diff
@@ -0,0 +1,56 @@
+        return ResponseEntity.ok().body(requestService.acceptRequest(id));
+    }
+
+    @PutMapping
```
аналогично

drewbal commented on src/main/java/school/faang/user_service/dto/MentorshipRequestDto.java:
```diff
@@ -0,0 +1,24 @@
+    @NotBlank(message = "Description most not be empty or blank.")
+    private String description;
+
+    private Long requesterId;
```
советую на эти тоже повесить аннотации, как минимум @Positive

DeemaP replied:
Здесь накинул

drewbal commented on src/main/java/school/faang/user_service/dto/RequestFilterDto.java:
```diff
@@ -0,0 +1,14 @@
+public class RequestFilterDto {
+    private String descriptionPattern;
+    private Long requesterId;
+    private Long receiverId;
```
Ну и сюда валидаций навешай

DeemaP replied:
Тоже накинул валидацию по длине описания и позитив

drewbal reviewed (commented):
Капелька советов по RestController

DeemaP committed:
Filter interface
refactor data naming.

DeemaP committed:
Merge branch 'refs/heads/Dragon-master-stream7' into feature-BJS2-40489

# Conflicts:
#	src/main/java/school/faang/user_service/service/UserService.java
#	src/main/java/school/faang/user_service/service/mentorship/MentorshipService.java

DeemaP committed:
Conflicts resolve.

DeemaP committed:
Merge branch 'refs/heads/Dragon-master-stream7' into feature-BJS2-40489

DeemaP committed:
Merge branch 'refs/heads/Dragon-master-stream7' into feature-BJS2-40489

# Conflicts:
#	src/test/java/school/faang/user_service/service/user/UserServiceTest.java
